### PR TITLE
Extract constants from api-gen to separate project and use it in entry-gen and IDE-Plugin

### PR DESCRIPTION
**Pull request by chippmann**
This extracts the contents of the `constants` and parts of the `extensions` package of the api-gen into its own gradle project `tools-common`.

This is done to commonize all hardcoded occurences of fqNames from godot and our own.

This also made a slight restructuring of our gradle project setup necessary as outlined on Discord. Quick recap:
The api-gen is a gradle plugin. As such it is a separate gradle project included in our main project through the gradle composit builds feature. As such it has no direct access to any other modules in our main project. But with this PR it needs access to the new `tools-common` project as it depends on it. As a result, the `tools-common` cannot just be a module, it needs to be a separate gradle project included in the build with gradles composite build feature (just like the api-gen).
Furthermore as the `tools-common` is now a dependency of both the entry-gen and the IDE plugin, it needs to be published. But our gradle publish plugin resides in `buildSrc` which is not accessible by `tools-common` as it is a separate gradle project. Thus our publish plugin was made a separate gradle project as well.
Furthermore as the `tools-common` is published, it also needs the proper versioning. This was also done through `buildSrc` which, as said previously` is also not accessible from `tools-common` as it is a separate gradle project. As a consequence, this part was extracted into its own gradle plugin called `godot-dependenices`.

To summarize:
- `api-gen` is still a gradle plugin and as such a separate gradle project included through composite build
- `tools-common` is a new separate gradle project included through composite build
- our publish plugin is extracted from `buildSrc` into its own gradle plugin project included through composite build
- the version definition and depdenency version declaration is moved from `buildSrc` into its own gradle plugin project included through composite build
- constants are moved from `api-gen` to `tools-common`
- all hardcoded occurences of fqNames in the entry-gen and IDE plugin are replaced with the constants from `tools-common`

**Note:** This PR does not address anything in regards to the trait system introduced by @piiertho. This will be looked at in a separate PR after this one as it's already large enough.

### CODE REVIEW

File: kt/api-generator/build.gradle.kts [hunk 1]
```diff
@@ -1,6 +1,10 @@
+import godot.dependencies.gradle.DependenciesVersions
+import godot.dependencies.gradle.fullGodotKotlinJvmVersion
+
 plugins {
     `kotlin-dsl`
     `java-gradle-plugin`
+    id("com.utopia-rise.godot-dependenices")
```
Comment (piiertho) [inline kt/api-generator/build.gradle.kts:7]: typo `dependenices` => `dependencies`

File: kt/api-generator/settings.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+
+pluginManagement {
+    resolutionStrategy.eachPlugin {
+        when (requested.id.id) {
+            "com.utopia-rise.godot-dependenices" -> useModule("com.utopia-rise:godot-dependencies-gradle-plugin:0.0.1")
```
Comment (piiertho) [inline kt/api-generator/settings.gradle.kts:5]: Same here

File: kt/tools-common/src/main/kotlin/godot/tools/common/constants/Packages.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+const val signalPackage = "godot.signals"
+const val godotApiPackage = "godot"
+const val godotCorePackage = "godot.core"
+const val godotMemoryPackage = "godot.core.memory"
+const val godotAnnotationPackage = "godot.annotation"
+const val godotUtilPackage = "godot.util"
+
+const val godotRegistrationPackage = "godot.registration"
+const val godotSignalsPackage = "godot.signals"
+const val godotEntryGeneratorBasePackage = "godot"
```
Comment (piiertho) [inline kt/tools-common/src/main/kotlin/godot/tools/common/constants/Packages.kt:12]: I would rename this to `godotEntryBasePackage`, this is misleading.

File: kt/entry-generation/godot-kotlin-symbol-processor/build.gradle.kts [hunk 4]
```diff
@@ -1,8 +1,9 @@
-import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import godot.dependencies.gradle.DependenciesVersions
 
 plugins {
     kotlin("jvm")
     id("com.utopia-rise.godot-publish")
+    id("com.utopia-rise.godot-dependenices")
```
Comment (piiertho) [inline kt/entry-generation/godot-kotlin-symbol-processor/build.gradle.kts:6]: Same typo here.

File: kt/godot-library/build.gradle.kts [hunk 5]
```diff
@@ -4,7 +4,8 @@ plugins {
     kotlin("jvm")
     id("com.utopia-rise.api-generator")
     id("com.utopia-rise.godot-publish")
-    id("com.github.johnrengelman.shadow") version DependenciesVersions.shadowJarPluginVersion
+    id("com.utopia-rise.godot-dependenices")
```
Comment (piiertho) [inline kt/godot-library/build.gradle.kts:7]: Same here.

File: kt/plugins/godot-gradle-plugin/build.gradle.kts [hunk 6]
```diff
@@ -1,10 +1,12 @@
+import godot.dependencies.gradle.DependenciesVersions
+
 plugins {
     kotlin("jvm")
     `maven-publish`
     `java-gradle-plugin`
     id("com.gradle.plugin-publish") version "0.14.0"
     id("com.utopia-rise.godot-publish")
-    id("org.ajoberstar.grgit")
+    id("com.utopia-rise.godot-dependenices")
```
Comment (piiertho) [inline kt/plugins/godot-gradle-plugin/build.gradle.kts:9]: Same here.

File: kt/plugins/godot-intellij-plugin/build.gradle.kts [hunk 7]
```diff
@@ -11,8 +13,7 @@ plugins {
     id("org.jetbrains.intellij") version "1.7.0"
     // gradle-changelog-plugin - read more: https://github.com/JetBrains/gradle-changelog-plugin
     id("org.jetbrains.changelog") version "1.3.1"
-    // detekt linter - read more: https://detekt.github.io/detekt/gradle.html
-    id("io.gitlab.arturbosch.detekt") version "1.20.0"
+    id("com.utopia-rise.godot-dependenices")
```
Comment (piiertho) [inline kt/plugins/godot-intellij-plugin/build.gradle.kts:16]: Same here.

File: kt/settings.gradle.kts [hunk 8]
```diff
@@ -1,16 +1,20 @@
+
 include("godot-library")
 
 pluginManagement {
     resolutionStrategy.eachPlugin {
-        if (requested.id.id == "com.utopia-rise.api-generator") {
-            useModule("com.utopia-rise:api-generator:0.0.1")
+        when(requested.id.id) {
+            "com.utopia-rise.godot-dependenices" -> useModule("com.utopia-rise:godot-dependencies-gradle-plugin:0.0.1")
```
Comment (piiertho) [inline kt/settings.gradle.kts:7]: Same here.

File: kt/tools-common/build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import godot.dependencies.gradle.DependenciesVersions
+import godot.dependencies.gradle.fullGodotKotlinJvmVersion
+
+plugins {
+    `kotlin-dsl`
+    id("org.ajoberstar.grgit") version "4.1.1"
+    id("com.utopia-rise.godot-publish")
+    id("com.utopia-rise.godot-dependenices")
```
Comment (piiertho) [inline kt/tools-common/build.gradle.kts:8]: Same here.

File: kt/tools-common/settings.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+
+pluginManagement {
+    resolutionStrategy.eachPlugin {
+        when (requested.id.id) {
+            "com.utopia-rise.godot-dependenices" -> useModule("com.utopia-rise:godot-dependencies-gradle-plugin:0.0.1")
```
Comment (piiertho) [inline kt/tools-common/settings.gradle.kts:5]: Same here.

File: kt/utils/godot-build-props/build.gradle.kts [hunk 11]
```diff
@@ -1,9 +1,11 @@
 import org.apache.tools.ant.filters.ReplaceTokens
+import godot.dependencies.gradle.DependenciesVersions
 
 plugins {
     kotlin("jvm")
     `maven-publish`
     id("com.utopia-rise.godot-publish")
+    id("com.utopia-rise.godot-dependenices")
```
Comment (piiertho) [inline kt/utils/godot-build-props/build.gradle.kts:8]: Same here.

**Review by CedNaru — APPROVED** [review]

**Review by piiertho — APPROVED** [review]

**Pull request merged**